### PR TITLE
AixPb: create hosts_file role for aix

### PR DESCRIPTION
**Pull request by Haroon-Khel**
Currently the aix playbook does not have a role/task which enters the correct entry in `/etc/hosts`, leading to improper entries added to this file; the correct entry being something that looks like

```
$public_ip $hostname.$domain $hostname

// For localhost
127.0.0.1   localhost  localhost.localdomain
```

So far this role is a direct copy of the `adopt_etc` role from the Unix playbook and is a draft. I will be making it suitable for aix, I just wanted to get the bulk of it down for now

### CODE REVIEW

File: ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#####################
+# Update /etc/hosts #
+#####################
+---
+- name: Update /etc/hosts file - IP FQDN hostname
+  lineinfile:
+    dest: /etc/hosts
+    regexp: "^(.*){{ ansible_hostname }}(.*)$"
+    line: "{{ ansible_default_ipv4.address }} {{ inventory_hostname }}.{{ Domain }} {{ inventory_hostname }}"
+    state: present
```
Comment (aixtools) [inline ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml:10]: * Note #1996 and use new identifier 
* Also, much prefer the shortname (ie, no domain) as first argument and hostname with domain as a comment (or at least last on line)

File: ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    - inventory_hostname != ansible_default_ipv4.address
+    - inventory_hostname != "localhost"
+  tags: hosts_file, adoptopenjdk
+
+- name: Update /etc/hosts file - IP FQDN hostname (Domain != "adoptopenjdk.net")
+  lineinfile:
+    dest: /etc/hosts
+    regexp: "^(.*){{ ansible_hostname }}(.*)$"
+    line: "{{ ansible_default_ipv4.address }} {{ ansible_fqdn }} {{ ansible_hostname }}"
+    state: present
```
Comment (aixtools) [inline ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml:22]: Likewise:
* Note #1996 and use new identifier
* Also, much prefer the shortname (ie, no domain) as first argument and hostname with domain as a comment (or at least last on line

File: ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    Or use the command line argument --extra-vars 'inventory_hostname=<realhostname>',
+    when running the playbook next time"
+  when: (inventory_hostname == ansible_default_ipv4.address) or (inventory_hostname == "localhost")
+  tags: hosts_file, hostname
+
+- name: Update /etc/hosts file - 127.0.0.1
+  lineinfile:
+    dest: /etc/hosts
+    regexp: "^(.*)127.0.0.1(.*)$"
+    line: "127.0.0.1   localhost  localhost.localdomain"
```
Comment (aixtools) [inline ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml:44]: This is what is customary on AIX (no .localdomain)
```
# grep localhost /etc/hosts
127.0.0.1               loopback localhost      # loopback (lo0) name/address
::1                     loopback localhost      # IPv6 loopback (lo0) name/address
```
* So, do not forget to verify IPv6 `::1` if you are going to look at localhost.

**Review by aixtools — CHANGES_REQUESTED** [review]
* see #1996
* shortname to have preference
* AIX localhost conventions are not Linux conventions
* Also note - the actual hostname might not be inventory_hostname and you may endup with two lines with the same IP address and name resolution is (officially) unspecified. iirc, the first entry in the hosts file is what gets returned.

**Comment by Haroon-Khel** [comment]
@aixtools Ive made some changes to the way the ipv4 and ipv6 loopback addresses are entered in /etc/hosts. What do you think? Tested on ojdk06 and build-osuosl-aix71-ppc64-1, the entries look like this

```
127.0.0.1   loopback  localhost
::1   loopback  localhost
```

**Comment by aixtools** [comment]
> 
> 
> @aixtools Ive made some changes to the way the ipv4 and ipv6 loopback addresses are entered in /etc/hosts. What do you think? Tested on ojdk06 and build-osuosl-aix71-ppc64-1, the entries look like this
> 
> ```
> 127.0.0.1   loopback  localhost
> ::1   loopback  localhost
> ```
> 
> External ip entries look like this
> 
> ```
> ipv4  hostname  hostname.domain
> ```

then lgtm

**Review by aixtools — COMMENTED** [review]
Missed this `placement` first time around - apologies.

Rather than create a new section (# 7) I would put it as part of the base configuration, maybe even changing the label for # 1 or # 2 to accommodate system settings.

iirc there are system settings hiding in the AIX filesystem scripts - and I would prefer, ultimately, a single role that covers aix_config - and additional host settings should fall into that category.

Ultimately, a review of the *UNIX* playbook roles will be the guide.

For now, it seems a bit out of place, and alone.

And, as I do not wish to have a long discussion re: changing this role name in the future to something more generic - permitting additional tasks (e.g., those not yet in a role).

Hmm: update: this was supposed to be a review of the AIX main file, sigh.

**Review by sxa — APPROVED** [review]
@aixtools has confirmed in slack he's OK with these changes as-is so I'm approving and merging

**Comment by aixtools** [comment]
Closed - but adding a comment for now:

Something is not quite right (yet):
```
TASK [hosts_file : Inventory is localhost message] **********************************************************************************************************************************************
fatal: [ojdk06]: FAILED! => {"msg": "The conditional check '(inventory_hostname == ansible_default_ipv4.address) or (inventory_hostname == \"localhost\")' failed. The error was: error while evaluating conditional ((inventory_hostname == ansible_default_ipv4.address) or (inventory_hostname == \"localhost\")): 'ansible_default_ipv4' is undefined\n\nThe error appears to be in '/home/aixtools/openjdk-infrastructure/ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/hosts_file/tasks/main.yml': line 30, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Inventory is localhost message\n  ^ here\n"}
```
* if I read above correctly: `ansible_default_ipv4' is undefined - so there is some ansible dependency in the playbook

**Pull request merged**